### PR TITLE
[TG-2755] Fix nondet-static option for static initializers

### PR DESCRIPTION
**Pull request by majakusber**
If `nondet-static` option is used, this makes sure that static variables that have a static initializer are not set to the value in the static initiliazer but nondet-initialized (by nondet-initializing all static variables of the class in `clinit_wrapper`).

### CODE REVIEW

**Comment by majakusber** [comment]
Tests and test-gen bump will follow.

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 1]
```diff
@@ -157,6 +157,11 @@ void java_bytecode_languaget::get_language_options(const cmdlinet &cmd)
   else
     java_cp_include_files=".*";
 
+  if(cmd.isset("nondet-static"))
+  {
+    nondet_static = true;
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: `generate_nondet_static_initializers`?

File: jbmc/src/java_bytecode/java_static_initializers.cpp [hunk 2]
```diff
@@ -184,9 +184,12 @@ gen_clinit_eqexpr(const exprt &expr, const clinit_statest state)
 /// \param class_name: name of the class to generate clinit wrapper calls for
 /// \param [out] init_body: appended with calls to clinit wrapper
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_static_initializers.cpp:188]: Update docs

File: jbmc/src/java_bytecode/java_static_initializers.cpp [hunk 3]
```diff
@@ -384,7 +421,10 @@ static void create_clinit_wrapper_symbols(
 /// \return the body of the static initialiser wrapper
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_static_initializers.cpp]: Update docs

File: jbmc/src/java_bytecode/java_static_initializers.cpp [hunk 4]
```diff
@@ -205,9 +208,43 @@ static void clinit_wrapper_do_recursive_calls(
...
+      {
+        if(
+          symbol.second.is_static_lifetime &&
+          symbol.second.type.get(ID_C_class) == class_name)
+        {
+          const symbol_exprt new_global_symbol = symbol.second.symbol_expr();
+
+          parameters.max_nonnull_tree_depth =
+            is_non_null_library_global(symbol.first)
+              ? object_factory_parameters.max_nonnull_tree_depth + 1
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_static_initializers.cpp]: Should this just be `std::max(1, object_factory_parameters.max_nonnull_tree_depth)`?

**Review by smowton — APPROVED** [review]
Looks good, some nitpickery

File: jbmc/src/java_bytecode/java_bytecode_language.cpp (hunk 1 above)
Comment (majakusber) [reply jbmc/src/java_bytecode/java_bytecode_language.cpp]: There are two reasons for this name - first, it copies the name of the flag itself which seems to be often the case in this function (it makes it easier to see the connection then?) and second, this option does not only generate nondet staic initializers. It nondet initializes everything in CPROVER_init and also goes over static initializers and uses nondet there.

**Review by peterschrammel — COMMENTED** [review]
@svorenova, there's a failed test: https://travis-ci.org/diffblue/cbmc/jobs/410887870#L7400

**Comment by majakusber** [comment]
@peterschrammel The test should be passing now.

**Comment by majakusber** [comment]
Pointer bump and regression tests can be found here: https://github.com/diffblue/test-gen/pull/2128

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 7ff9a8a3394eaf313ea1c98a5be84110ffdd5898).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80782147
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by peterschrammel — APPROVED** [review]
It would be good to have at least one JBMC test as well.

**Comment by majakusber** [comment]
@peterschrammel It seems `nondet-static` is not an option in jbmc. Is there a particular reason for it? If not then I can add it.

**Comment by peterschrammel** [comment]
>  can add it.

Yes, please.

**Comment by majakusber** [comment]
@smowton @peterschrammel I added the nondet-static option to JBMC and a test for it. I also changed the solution a bit - see the clinit_wrapper cont. commit - to make sure that `static final` fields get handled correctly.

File: jbmc/src/java_bytecode/java_static_initializers.cpp [hunk 5]
```diff
@@ -209,6 +220,43 @@ static void clinit_wrapper_do_recursive_calls(
...
+        {
+          const symbol_exprt new_global_symbol = symbol.second.symbol_expr();
+
+          parameters.max_nonnull_tree_depth =
+            is_non_null_library_global(symbol.first)
+              ? std::max(
+                  size_t(1), object_factory_parameters.max_nonnull_tree_depth)
+              : object_factory_parameters.max_nonnull_tree_depth;
+
+          gen_nondet_init(
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_static_initializers.cpp]: Careful -- gen-nondet-init mutates the symbol table, which invalidates iterators over the symbol table. You must first pick which identifiers to mess with, then retrieve and run gen-nondet-init for each in turn.

**Review by smowton — CHANGES_REQUESTED** [review]

File: jbmc/src/java_bytecode/java_static_initializers.cpp (hunk 5 above)
Comment (majakusber) [reply jbmc/src/java_bytecode/java_static_initializers.cpp]: Done :)

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2376b329fa3526f82317d16563d76cc74d01310c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81038517

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 8762296d09f2eb8205c3d91a3a98cc912c4f7f40).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81038249

**Comment by majakusber** [comment]
Squashed some commits, waiting for the final round of CI.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 86f51f54bf3097746fcabc0b2c92b94cf3bd5567).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81123619

**Comment by majakusber** [comment]
Extended the regression test with various combinations of public/private/final fields, initialization with a function call, using models, fields in inner classes...

**Comment by majakusber** [comment]
Test-gen pointer bump: https://github.com/diffblue/test-gen/pull/2195

**Comment by majakusber** [comment]
@smowton @peterschrammel Can you please have a quick second look at the last 4 commits? The tests have been extended, and I added the last three small commits related to this.

**Comment by smowton** [comment]
Looks good

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**